File: leetcode-383-Ransom_Note/leetcode-383-Ransom_Note.cpp
/* -- Id = 383 -- */

#include <cstdlib>
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

bool canConstruct(string ransomNote, string magazine) {
    unordered_map<char, size_t> m;

    for (auto &i : ransomNote)
    {
        if(m.find(i) == m.end())
        {
            m[i] = 0;
        }
        ++m[i];
    }

    for (auto &i : magazine)
    {
        if(m.find(i) != m.end())
        {
            --m[i];
            if(m[i] <= 0)
            {
                m.erase(i);
            }
        }
    }
    
    if (m.empty())
        return true;

    return false;
}

int main(int argc, char const *argv[])
{
    cout << canConstruct("aa", "aab") << endl;
    return EXIT_SUCCESS;
}

File: leetcode-383-Ransom_Note/leetcode-383-Ransom_Note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        letters_count: dict = {}

        for c in magazine:
            if c not in letters_count:
                letters_count[c] = 0
            letters_count[c]+=1
        # --------
        for c in ransomNote:
            if c in letters_count:
                letters_count[c]-=1
                if letters_count[c]<0:
                    return False
            else:
                return False
        return True


if __name__ == "__main__":
    ransomNote: str = "a"
    magazine: str = "b"

    res: bool = Solution().canConstruct(ransomNote, magazine)
    print(res)
